### PR TITLE
api: asset: delete & restore

### PR DESCRIPTION
**Pull request by gioelecerati**
**What does this pull request do? Explain your changes. (required)**

**Specific updates (required)**

**How did you test each of these updates (required)**

**Does this pull request close any open issues?**

**Screenshots (optional)**

**Checklist**

- [ ] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: packages/api/src/controllers/asset.ts [hunk 1]
```diff
@@ -706,6 +719,14 @@ app.get("/", authorizer({}), async (req, res) => {
     res.links({ next: makeNextHREF(req, newCursor) });
   }
 
+  if (req.user.admin && deleting) {
+    output = output.filter((asset) => asset.status.phase == "deleting");
+    output = output.filter(
+      (asset) =>
+        asset.deletedAt && Date.now() > asset.deletedAt + DELETE_ASSET_DELAY
+    );
+  }
```
Comment (victorges) [inline packages/api/src/controllers/asset.ts]: **Blocking:** We must filter assets on the SQL query instead, [like this](https://github.com/livepeer/studio/pull/2092/files#diff-ed7a14d016bf3f1e0d94b5823604fde2f1dba33ce1376c00c1237dfa3ed3775bL647), not filtering the DB response in memory.

File: packages/api/src/controllers/asset.ts [hunk 2]
```diff
@@ -1089,11 +1110,64 @@ app.delete("/:id", authorizer({}), async (req, res) => {
...
+  if (asset.status.phase === "deleting" || asset.deleted) {
+    throw new BadRequestError(`Asset is already deleted`);
+  }
+
   await req.taskScheduler.deleteAsset(asset);
   res.status(204);
   res.end();
 });
 
+app.put("/:id/restore", authorizer({}), async (req, res) => {
```
Comment (victorges) [inline packages/api/src/controllers/asset.ts]: I believe this should be a `POST` which AFAIK is the general fallback in REST for an "action" on a resource, which is what being triggered here.

File: packages/api/src/controllers/asset.ts [hunk 3]
```diff
@@ -1089,11 +1110,64 @@ app.delete("/:id", authorizer({}), async (req, res) => {
...
+  const { id } = req.params;
+  const asset = await db.asset.get(id);
+
+  if (!asset) {
+    throw new NotFoundError(`Asset not found`);
```
Comment (victorges) [inline packages/api/src/controllers/asset.ts]: I believe our errors are generally `lower case`, let's keep it consistent (other occurrences in this PR)

File: packages/api/src/task/scheduler.ts [hunk 4]
```diff
@@ -539,10 +539,44 @@ export class TaskScheduler {
       if (typeof asset === "string") {
         asset = await db.asset.get(asset);
       }
+
+      let phase = asset.status?.phase;
+      // Prevent bump of updatedAt if phase isn't getting updated
+      let updatedAt = asset.status?.updatedAt;
+
+      if (phase == "ready") {
```
Comment (victorges) [inline packages/api/src/task/scheduler.ts]: You gotta remember to use triple equals in JS, double equal can do weird stuff sometimes. As a rule of thumb, **always** use triple equals unless you have a reason not to.
```suggestion
      if (phase === "ready") {
```

File: packages/api/src/task/scheduler.ts [hunk 5]
```diff
@@ -539,10 +539,44 @@ export class TaskScheduler {
...
+        // If the asset is ready, we need to schedule deletion
```
Comment (victorges) [inline packages/api/src/task/scheduler.ts:548]: I wonder if we want to trigger this for failed assets too. Maybe there are left overs?

File: packages/api/src/task/scheduler.ts [hunk 6]
```diff
@@ -539,10 +539,44 @@ export class TaskScheduler {
...
+    }
+  }
+
+  async restoreAsset(asset: string | Asset) {
+    try {
+      if (typeof asset === "string") {
+        asset = await db.asset.get(asset);
+      }
+
+      await this.updateAsset(asset, {
```
Comment (victorges) [inline packages/api/src/task/scheduler.ts:573]: Will this send any webhooks as well?

**Review by victorges — CHANGES_REQUESTED** [review]
Requesting changes because filtering assets in memory is a no go. The other comments are optional.

File: packages/api/src/task/scheduler.ts (hunk 6 above)
Comment (gioelecerati) [reply packages/api/src/task/scheduler.ts:573]: we could send an asset.restored event, but I don't think we need it
restoring a deleted asset is potentially a rare occasion

**Review by victorges — APPROVED** [review]
LGTM

**Pull request merged**